project_restorer/src/main/scala/PlutoCoreMessageProcessor.scala

```diff
@@ -81,51 +131,91 @@ class PlutoCoreMessageProcessor(mxsConfig: MatrixStoreConfig)(implicit mat: Mate
     case _ => false
   }
 
-  def getNearlineResults(projectId: Int) = {
+  def getNearlineResults(projectId: Int): Future[Either[String, Seq[OnlineOutputMessage]]] =
     matrixStoreBuilder.withVaultFuture(mxsConfig.nearlineVaultId) { vault =>
       searchAssociatedNearlineMedia(projectId, vault).map(Right.apply)
     }
-  }
 
-  def getOnlineResults(projectId: Int) = {
-    searchAssociatedOnlineMedia(projectId, vidispineCommunicator).map(Right.apply)
+  def getOnlineResults(projectId: Int): Future[Right[Nothing, Seq[OnlineOutputMessage]]] = {
+    searchAssociatedOnlineMedia(vidispineCommunicator, projectId).map(Right.apply)
   }
 
   def handleUpdateMessage(updateMessage: ProjectUpdateMessage, framework: MessageProcessingFramework): Future[Either[String, MessageProcessorReturnValue]] =
     updateMessage.status match {
       case status if statusesMediaNotRequired.contains(status) =>
-        Future
-          .sequence(Seq(getNearlineResults(updateMessage.id), getOnlineResults(updateMessage.id)))
-          .map(allResults => processResults(allResults, RoutingKeys.MediaNotRequired, framework, updateMessage.id, updateMessage.status))
+
+        getOnlineResults(updateMessage.id).flatMap({
+          case Right(onlineResults: Seq[OnlineOutputMessage]) =>
+
+            // We don't want to send any media_not_required.nearline messages if status is Held
```

Comment:
It would be nice to separate out the parts here into seperate functions, which we could then iterate over

e.g.
```scala
if(updateMessage.status==EntryStatus.Held.toString) {
                 logger.debug(s"Status of project ${updateMessage.id} is ${EntryStatus.Held}, we won't consider any nearline files")
                Future(Right(Seq()))
} else {
  for {
    nearlineResults <- getNearlineResults(updateMessage.id)
    crosslinkstatusItemTuples <- enhanceOnlineResultsWithCrosslinkStatus(onlineResults)
    _ = logOutBasis(onlineResults, crosslinkstatusItemTuples.size)
    (filteredOnline, filteredNearline) = crosslinkFilters(crosslinkstatusItemTuples)
    result <- processResults(filteredNearline, filteredOnline, RoutingKeys.MediaNotRequired, framework, updateMessage.id, updateMessage.status)
  } yield result
}

....

private def logOutBasis(onlineResults:xxxxx, itemCount:Int):Unit = {
        if (itemCount < MAX_ITEMS_TO_LOG_INDIVIDUALLY) {
          logger.debug(s"Basis for filtering items for project ${updateMessage.id} (online items as vsItemId|nearlineId|path|crosslinkedIds->SendRemoveActionTarget) " +
            crosslinkstatusItemTuples.map(
              t =>
                s"${t._2.vidispineItemId.getOrElse("<no vs ID>")}|" +
                  s"${t._2.nearlineId.getOrElse("<no online ID>")}|" +
                  s"${t._2.originalFilePath.getOrElse("<no path>")}|" +
                  s"${getCrosslinkProjectIds(t._2)}->" +
                  s"${t._1}"
            ).mkString(", "))
        } else {
          logger.debug(s"${itemCount} crosslink/onlineItem tuples, too many to list all in log")
        }
}

private def crosslinkFilterOnline(crosslinkStatusItemTuples:xxxx) = {
   ...[all of the val =filter statement lines]....
   logPreAndPostCrosslinkFiltering(....)
   (filteredOnline, filteredNearline)
}
```

Reply:
That said, I am making a few probably incorrect assumptions about the return types of functions for the for-comprehension.

Remember that `<-` in the for-comprehansion means `map` so the types must be mappable, e.g. if the first line returns `Future[Either[X,Y]]` then subsequent lines should also return `Future[T]`.

`=` does not mean map and retains its normal meaning

So, if `getNearlineResults` returns a `Future[Either[X,Y]]` but nothing else is a future, you might want to tweak the other functions called to also return `Either[X,Y]` and map over that instead:

```scala
  getNearlineResults.map(nearlineResults=>
    for {
      crosslinkstatusItemTuples <- enhanceOnlineResultsWithCrosslinkStatus(onlineResults) //assumes that `enhanceOnlineResultsWithCrosslinkStatus` returns an `Either` now
.
.
.
```